components/list/src/patterns.rs

```diff
@@ -0,0 +1,288 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::provider::*;
+use crate::ListLength;
+#[cfg(feature = "datagen")]
+use alloc::borrow::Cow;
+#[cfg(feature = "datagen")]
+use icu_provider::DataError;
+use regex_automata::dfa::Automaton;
+use writeable::{LengthHint, Writeable};
+
+impl<'data> ListFormatterPatternsV1<'data> {
+    /// Creates a new [`ListFormatterPatternsV1`] from the given patterns. Fails if any pattern is invalid.
+    ///
+    /// See [`ListJoinerPattern::from_str`]. `allow_prefix` will be true for `pair` and `end` patterns,
+    /// `allow_suffix` for `start` and `pair` patterns.
+    #[cfg(feature = "datagen")]
+    pub fn try_new(
```

Comment:
suggestion: should these just be twelve parameters?

Reply:
Yes, but that'd be a breaking change now.

Reply:
ah